spec/date_range_spec.rb

```diff
@@ -0,0 +1,74 @@
+require_relative 'spec_helper'
+
+describe "DateRange Class" do
+  before do
+    @date1 = Date.parse('2018-09-05')
+    @date2 = Date.parse('2018-09-09')
+    @date3 = Date.parse('2018-09-10')
+    @date4 = Date.parse('2018-09-11')
+
+    @date_range1 = Hotel::DateRange.new(@date1, @date2)
+    @date_range2 = Hotel::DateRange.new(@date1, @date3)
+    @date_range3 = Hotel::DateRange.new(@date2, @date3)
+    @date_range4 = Hotel::DateRange.new(@date3, @date4)
+  end
+
+  describe "Date Range initation " do
+    it "is an instance of Date Range" do
+      expect(@date_range1).must_be_kind_of Hotel::DateRange
+    end
+
+    it "passes Dates as arguments" do
+      expect(@date_range1.check_in).must_be_kind_of Date
+      expect(@date_range2.check_out).must_be_kind_of Date
+    end
+
+    it "raises a StandardError if invalid date range is provided" do
+      expect{ Hotel::DateRange.new(@date1, @date1) }.must_raise StandardError
+      expect{ Hotel::DateRange.new(@date3, @date1) }.must_raise StandardError
+    end
+    it "raises an ArgumentError if user do not put in a Date" do
+      expect{ Hotel::DateRange.new(@date1, '12/08/23') }.must_raise ArgumentError
+      expect{ Hotel::DateRange.new('12/06/30', @date1) }.must_raise ArgumentError
+      expect{ Hotel::DateRange.new('12/06/30', '12/08/23') }.must_raise ArgumentError
+    end
+
+    it "creates an array of Dates between check_in and check_out & excluding check_out" do
+      expect(@date_range3.dates_booked).must_be_kind_of Array
+      expect(@date_range3.dates_booked.length).must_equal 1
+      expect(@date_range3.dates_booked[0]).must_equal @date2
+      expect(@date_range2.dates_booked.length).must_equal 5
+      expect(@date_range2.dates_booked[0]).must_equal @date1
+    end
+  end
+
+  describe "overlaps? method to check if ranges overlap" do
+    it "returns true if date ranges are the same" do
+      expect(@date_range1.overlaps?(@date_range1)).must_equal true
+    end
+
+    it "returns true if date ranges overlaps in the front" do 
+      expect(@date_range1.overlaps?(@date_range2)).must_equal true
+    end
+
+    it "returns true if date ranges overlaps in the back" do
+      expect(@date_range2.overlaps?(@date_range1)).must_equal true
+    end
+
+    it "returns true if one range is completely contained in another range" do
+      expect(@date_range2.overlaps?(@date_range3)).must_equal true
+    end
+
+    it "returns false if one range is completely after" do
+      expect(@date_range1.overlaps?(@date_range4)).must_equal false
+    end
+
+    it "returns false if one range is completely before" do
+      expect(@date_range4.overlaps?(@date_range1)).must_equal false
+    end
+    it "returns false if a range starts on another's check in date" do
+      expect(@date_range1.overlaps?(@date_range3)).must_equal false
+      expect(@date_range2.overlaps?(@date_range4)).must_equal false
+    end
+  end
+end
```

Comment:
I love this DateRange class because it isolated the overlapping date logic to this one class, which is complex and tested in isolation... Beautiful :* truly inspiring :*